asmcomp/cmmgen.ml

```diff
@@ -1107,49 +1118,60 @@ let alloc_matches_boxed_int bi ~hdr ~ops =
         && String.equal sym caml_int64_ops
   | (Pnativeint | Pint32 | Pint64), _, _ -> false
 
-let rec unbox_int bi arg dbg =
-  match arg with
-    Cop(Calloc, [hdr; ops; Cop(Clsl, [contents; Cconst_int (32, _)], dbg')],
-      _dbg)
-    when bi = Pint32 && size_int = 8 && big_endian
-      && alloc_matches_boxed_int bi ~hdr ~ops ->
-      (* Force sign-extension of low 32 bits *)
-      Cop(Casr, [Cop(Clsl, [contents; Cconst_int (32, dbg)], dbg');
-        Cconst_int (32, dbg)],
-        dbg)
-  | Cop(Calloc, [hdr; ops; contents], _dbg)
-    when bi = Pint32 && size_int = 8 && not big_endian
-      && alloc_matches_boxed_int bi ~hdr ~ops ->
-      (* Force sign-extension of low 32 bits *)
-      Cop(Casr, [Cop(Clsl, [contents; Cconst_int (32, dbg)], dbg);
-        Cconst_int (32, dbg)],
-        dbg)
-  | Cop(Calloc, [hdr; ops; contents], _dbg)
-    when alloc_matches_boxed_int bi ~hdr ~ops ->
-      contents
-  | Clet(id, exp, body) -> Clet(id, exp, unbox_int bi body dbg)
-  | Cifthenelse(cond, ifso_dbg, e1, ifnot_dbg, e2, dbg) ->
-      Cifthenelse(cond,
-        ifso_dbg, unbox_int bi e1 ifso_dbg,
-        ifnot_dbg, unbox_int bi e2 ifnot_dbg,
-        dbg)
-  | Csequence(e1, e2) -> Csequence(e1, unbox_int bi e2 dbg)
-  | Cswitch(e, tbl, el, dbg') ->
-      Cswitch(e, tbl,
-        Array.map (fun (e, dbg) -> unbox_int bi e dbg, dbg) el,
-        dbg')
-  | Ccatch(rec_flag, handlers, body) ->
-      map_ccatch (fun e -> unbox_int bi e dbg) rec_flag handlers body
-  | Ctrywith(e1, id, e2, handler_dbg) ->
-      Ctrywith(unbox_int bi e1 dbg, id,
-        unbox_int bi e2 handler_dbg, handler_dbg)
-  | _ ->
-      if size_int = 4 && bi = Pint64 then
-        split_int64_for_32bit_target arg dbg
-      else
-        Cop(
-          Cload((if bi = Pint32 then Thirtytwo_signed else Word_int), Mutable),
-          [Cop(Cadda, [arg; Cconst_int (size_addr, dbg)], dbg)], dbg)
+let unbox_int dbg bi =
+  let default arg =
+    if size_int = 4 && bi = Pint64 then
+      split_int64_for_32bit_target arg dbg
+    else
+      Cop(
+        Cload((if bi = Pint32 then Thirtytwo_signed else Word_int),
+              Immutable),
```

Comment:
Note the change from `Mutable` to `Immutable`, unrelated to this PR, but correct AFAICT.